hazelcast/src/main/java/com/hazelcast/concurrent/atomicreference/AtomicReferenceService.java

```diff
@@ -174,4 +196,99 @@ public String getQuorumName(String name) {
+
+        for (Map.Entry<String, AtomicReferenceContainer> entry : containers.entrySet()) {
+            AtomicReferenceContainer container = entry.getValue();
+            if (!(getMergePolicy(container) instanceof DiscardMergePolicy)) {
+                String name = entry.getKey();
+                int partitionId = partitionService.getPartitionId(StringPartitioningStrategy.getPartitionKey(name));
+                if (partitionService.getPartition(partitionId, false).isLocal()) {
+                    // add your owned values to the map so they will be merged
+                    List<AtomicReferenceContainer> containerList = containerMap.get(partitionId);
+                    if (containerList == null) {
+                        containerList = new ArrayList<AtomicReferenceContainer>(containers.size());
+                        containerMap.put(partitionId, containerList);
+                    }
+                    containerList.add(container);
+                }
+            }
+        }
+        containers.clear();
+
+        return new Merger(containerMap);
+    }
+
+    private SplitBrainMergePolicy getMergePolicy(AtomicReferenceContainer container) {
+        String mergePolicyName = container.getConfig().getMergePolicyConfig().getPolicy();
+        return mergePolicyProvider.getMergePolicy(mergePolicyName);
+    }
+
+    private class Merger implements Runnable {
+
+        private static final int TIMEOUT_FACTOR = 500;
+
+        private final Map<Integer, List<AtomicReferenceContainer>> containerMap;
+
+        Merger(Map<Integer, List<AtomicReferenceContainer>> containerMap) {
+            this.containerMap = containerMap;
+        }
+
+        @Override
+        public void run() {
+            final ILogger logger = nodeEngine.getLogger(AtomicReferenceService.class);
+            final Semaphore semaphore = new Semaphore(0);
+
+            ExecutionCallback<Object> mergeCallback = new ExecutionCallback<Object>() {
+                @Override
+                public void onResponse(Object response) {
+                    semaphore.release(1);
+                }
+
+                @Override
+                public void onFailure(Throwable t) {
+                    logger.warning("Error while running merge operation: " + t.getMessage());
+                    semaphore.release(1);
+                }
+            };
+
+            // we cannot merge into a 3.9 cluster, since not all members may understand the MergeOperation
+            if (nodeEngine.getClusterService().getClusterVersion().isLessThan(Versions.V3_10)) {
+                logger.info("Cluster needs to run version " + Versions.V3_10 + " to merge AtomicReference instances");
+                return;
+            }
+
+            int valueCount = 0;
+            for (Map.Entry<Integer, List<AtomicReferenceContainer>> entry : containerMap.entrySet()) {
+                // TODO: add batching via an OperationFactory (see PutAllPartitionAwareOperationFactory)
```

Comment:
Will this be done for this PR or later?

Reply:
I just changed this TODO for `AtomicLong` and `AtomicReference`:
```
// TODO: add batching (which is a bit complex, since AtomicReference is a single-value data structure,
// so we need an operation for multiple AtomicReference instances, which doesn't exist so far)
```
If I'm not wrong, this is a quite complex task. Of course it would give a very nice speed-up, especially if you have a high number of these data structures. But since we're also developing a CP variant of this, I'm thinking of skipping the batching implementation for these data structures at all, due to the complexity of operation which needs to be introduced.

So far I would move this to the end of the dev sprint or 3.11.